### ISSUE TITLE
linker errors for static constexpr class members when compiled with c++17 in ndk 16 and 17

### ISSUE THREAD

**Issue by martinbonaciugome** [android-ndk/ndk#651]
#### Description

static constexpr class members get linker errors in ndk16b and ndk17-beta1 wich c++17 turned on.

there are many such linker errors, but this one example (i havent tested this exact reduced code though but it fails in a larger project).

`/// softfp.h
namespace xxx
{
  struct softfp
  {
    unsigned x;
  };
}
namespace std
{
  template<>
  struct numeric_limits<xxx::softfp>
  {
    static constexpr bool has_quiet_NaN = false;
    static constexpr bool has_signaling_NaN = false;
    static constexpr float_denorm_style has_denorm = denorm_absent;
    static constexpr int max_digits10 = 9;
    static constexpr int digits10 = 6;

    static constexpr xxx::softfp min() noexcept
    {
      return xxx::softfp{0x00800000};
    }
    static constexpr xxx::softfp lowest() noexcept
    {
      return xxx::softfp{0xff7fffff};
    }
    static constexpr xxx::softfp max() noexcept
    {
      return xxx::softfp{0x7f7fffff};
    }
    static constexpr xxx::softfp epsilon() noexcept
    {
      return xxx::softfp{0x33d6bf95};
    }
    static constexpr iugo::softfp round_error() noexcept
    {
      return xxx::softfp{0x3f000000};
    }
    static constexpr xxx::softfp infinity() noexcept
    {
      return xxx::softfp{0x7f800000};
    }
    static constexpr xxx::softfp quiet_NaN() noexcept
    {
      return xxx::softfp{0};
    }
    static constexpr xxx::softfp signaling_NaN() noexcept
    {
      return xxx::softfp{0};
    }
    static constexpr xxx::softfp denorm_min() noexcept
    {
      return xxx::softfp{0};
    }
  };
}
namespace xxx
{
  template<>
  struct numeric_limits<softfp>: std::numeric_limits<softfp>
  {
    static constexpr softfp min = std::numeric_limits<softfp>::min();
    static constexpr softfp lowest = std::numeric_limits<softfp>::lowest();
    static constexpr softfp max = std::numeric_limits<softfp>::max();
    static constexpr softfp epsilon = std::numeric_limits<softfp>::epsilon();
    static constexpr softfp round_error = std::numeric_limits<softfp>::round_error();
    static constexpr softfp infinity = std::numeric_limits<softfp>::infinity();
    static constexpr softfp quiet_NaN = std::numeric_limits<softfp>::quiet_NaN();
    static constexpr softfp signaling_NaN = std::numeric_limits<softfp>::signaling_NaN();
    static constexpr softfp denorm_min = std::numeric_limits<softfp>::denorm_min();
    
    using int_type = unsigned;
    static constexpr unsigned mant_mask = 0x007fffff;
    static constexpr unsigned exp_one_mask = 0x00800000;
    static constexpr unsigned exp_zero = 0x3f800000;
  };
}

/// softfp.cpp
namespace iugo
{
  constexpr softfp::int8 softfp::float_rounding_mode;
  constexpr softfp::int8 softfp::float_detect_tininess;
  constexpr softfp::float32bit softfp::float32_default_nan;
  constexpr softfp::float32bit softfp::float32_inf;

  constexpr softfp numeric_limits<softfp>::min;
  constexpr softfp numeric_limits<softfp>::lowest;
  constexpr softfp numeric_limits<softfp>::max;
  constexpr softfp numeric_limits<softfp>::epsilon;
  constexpr softfp numeric_limits<softfp>::round_error;
  constexpr softfp numeric_limits<softfp>::infinity;
  constexpr softfp numeric_limits<softfp>::quiet_NaN;
  constexpr softfp numeric_limits<softfp>::signaling_NaN;
  constexpr softfp numeric_limits<softfp>::denorm_min;
}

/// another.cpp
void main()
{
  cout << xxx::numeric_limits<xxx::softfp>::min.x;
}

///
linker:
undefined references during linking to xxx::numeric_limits<xxx::softfp>::min
`

#### Environment Details

Not all of these will be relevant to every bug, but please provide as much
information as you can.

* NDK Version: r16b or r17-beta1
* Build sytem: cmake
* Host OS: Windows
* Compiler: clang

**Comment by martinbonaciugome** [comment]
sorry then the code will give undefined references during linking to xxx::numeric_limitsxxx::softfp::min

**Comment by DanAlbert** [comment]
Please upload a test case that compiles. If I have to fix it to compile I'm not going to look as soon.

You also deleted a really important detail from the template: which STL are you using?

**Comment by martinbonaciugome** [comment]
ok i got a reduced form of this, and its not stl related.

a.h
`
namespace a
{
  template<class T>
  struct foo;
}
namespace b
{
  namespace a
  {
    using namespace ::a;
  }

  template<class T>
  struct foo;

  struct x
  {
    int y;

    x() = default;
    explicit constexpr x(int _y) noexcept
      : y{_y}
    {
    }
  };
}
namespace a
{
  template<>
  struct foo<b::x>
  {
    static constexpr b::x max() noexcept
    {
      return b::x{1};
    }
  };
}
namespace b
{
  template<>
  struct foo<x>
  {
    static constexpr auto max = a::foo<x>::max();
  };
}
namespace c
{
  void do_stuff();
}
`
a.cpp
`
#include "a.h"
#include <iostream>
#include <type_traits>
namespace b
{
  constexpr x foo<x>::max;
}
namespace c
{
  template<class T, typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
  constexpr int stuff(T const& rhs) noexcept
  {
    return rhs;
  }
  void do_stuff()
  {
    std::cout << stuff(b::foo<b::x>::max.y);
  }
}
`
main.cpp
`
#include "a.h"
int main()
{
  c::do_stuff();
  return 0;
}
`

a.cpp is in a static lib, and im building an "so". im not sure if that makes a difference. the problem seems to be because of the call to "stuff". calling "std::cout << b::foo<b::x>::max.y" doesn't produce this linker error. compiling with c++14 its fine. also in apples version of clang this works with c++17. plus there is something wrong with code formatting (using ``) in this post.

**Comment by rprichard** [comment]
More reduced test case:

```c++
template <int I>
struct Foo {};

template<> struct Foo<0> {
  static constexpr int bar = 42;
};

// This namespace-scope declaration was required in C++14. It's deprecated in C++17,
// but still allowed.
constexpr int Foo<0>::bar;

const int &f() {
  return Foo<0>::bar;
}

int main() {
  return f();
}
```

Building an executable with a standalone ARM android-16 compiler:

```
$ ~/android-ndk-r17-beta1/stand-arm-16/bin/clang++ test.cpp -Os -std=c++17
/tmp/test-0f3c7c.o:test.cpp:function f(): error: undefined reference to 'Foo<0>::bar'
/tmp/test-0f3c7c.o:test.cpp:function main: error: undefined reference to 'Foo<0>::bar'
```

See http://en.cppreference.com/w/cpp/language/static#Constant_static_members, and in particular:

> If a static data member is declared constexpr, it is implicitly inline and does not need to be redeclared at namespace scope. This redeclaration without an initializer (formerly required as shown above) is still permitted, but is deprecated. (since C++17)

When the code above is compiled with C++17, the `Foo<0>::bar` constexpr data member is an inline variable, and there is an upstream Clang bug where symbols for inline variables of template specializations aren't output. (https://bugs.llvm.org/show_bug.cgi?id=35027)

The upstream bug is fixed already, but it hasn't been included in an NDK compiler yet (either r16 or r17-beta1). This issue is probably a duplicate of https://github.com/android-ndk/ndk/issues/591. The test cases in both issues are fixed in trunk Clang, but not in NDK Clang.

**Comment by martinbonaciugome** [comment]
thanks for looking into this. will the latest clang with the fix be in for r17 release?

**Comment by rprichard** [comment]
Right now, it looks like r18 is more likely.

**Comment by rprichard** [comment]
This looks fixed in the r18 canary clang (4751641 based on r328903).
